pretyping/cases.ml

```diff
@@ -2150,12 +2150,40 @@ let prepare_predicate ?loc ~program_mode typing_fun env sigma tomatchs arsign ty
       (* We extract the signature of the arity *)
       let hypnaming = RenameExistingBut (VarSet.variables (Global.env ())) in
       let building_arsign,envar = List.fold_right_map (push_rel_context ~hypnaming sigma) arsign env in
-      (* We put a type constraint on the predicate so that one
-         branch type-checked first does not lead to a lower type than
-         another branch; we take into account the possible elimination
-         constraints on the predicate *)
-      let sigma, rtnsort = fresh_sort_in_family sigma (expected_elimination_sort !!env tomatchs) in
+      let sigma, rtnsort = Evd.new_sort_variable univ_flexible sigma in
       let sigma, predcclj = typing_fun (Some (mkSort rtnsort)) envar sigma rtntyp in
+      (* We take into account the possible elimination constraints on the
+        predicate. When there is an elimination constraint and the predicate is
+        underspecified, i.e. a QSort, we make a non-canonical choice for the
+        return type. Incompatible constraints are ignored and handled later
+        when typing the patter-matching. *)
+      let sigma = match expected_elimination_sort !!env tomatchs with
+      | InType ->
+        (* Not squashed, no constraints *)
+        sigma
+      | InProp ->
+        (* Squashed inductive in Prop, return sort must be Prop or SProp *)
```

Comment:
Not sure to be able to find an optimal formulation but here is a proposition:
```suggestion
        (* At least one term to match cannot be eliminated to sorts higher than Prop or SProp *)
```
and similarly below

Reply:
The presentation of sorts as a totally ordered lattice should be removed, because it's not accurate anymore since the introduction of SProp. We are confusing cumulative subtyping with eliminability, which are two different notions. It is going to become less and less true if we move to a multi-universe model. I'd advocate against using phrasing like "higher" in this kind of documentation, and I'm also actually going to submit PRs tweaking the API for eliminability that will remove any reference to an order.

Reply:
So, maybe: "At least one term to match in an inductive squashed in Prop: return sort must be Prop or SProp"?

Reply:
But then this is true for all branches since we could have chained eliminations over an arbitrarily long list of arguments.

Reply:
Then, how to say it? "Elimination constraints for the terms to match leave only Prop and SProp as possible type for the return sort"

(Maybe this is what "squashed" was intended to mean in your original comment.)

Anyway, I'm fine as it.

Reply:
> It is going to become less and less true if we move to a multi-universe model.

Just to be sure I understand what you have in mind: When we currently write `forall A:s, B`, we somehow implicitly write `forall s' ∈ {s'' | s'' ≤ s}, forall A:s', B`, where the expression `{s'' | s'' ≤ s}` describes a finite set. Is it a correct understanding that in the multi-universe model, the description of this finite set is instead in extension, or at least using a more liberal syntax than just of the form of a downards set?

Reply:
The details are not fully fleshed out, but in any case quantifying over quality will never incur a complex constraint system. It will be a pair of 1. a finite matrix of ground instantiations and 2. a generic scheme without any reference to any order.